### DISCUSSION TITLE
How to serialize and deserialize `PreparsedDocumentEntry` across the wire

### DISCUSSION THREAD

**Discussion by softprops** [graphql-java/graphql-java#2775] [Q&A] [answered]
In the latest release, at this time 18.0, there was a note the [changelog](https://github.com/graphql-java/graphql-java/releases/tag/v18.0) about [PreparsedDocumentProvider](https://www.javadoc.io/doc/com.graphql-java/graphql-java/18.0/graphql/execution/preparsed/PreparsedDocumentProvider.html)

> The graphql.execution.preparsed.PreparsedDocumentProvider has been updated to use a CompletableFuture<PreparsedDocumentEntry> getDocumentAsync method. This allows you to go to a remote cache in an asynchronous non blocking manner to get cached documents. This should have always been modelled in this manner.

I think think this is neat but what would be the recommended way for one to serialize and deserialize `PreparsedDocumentEntry` instances across a wire?

My key is the string query which I can easily hash to a key but I wasn't sure about the `PreparsedDocumentEntry` object itself.

**Comment by bbakerman** [comment, marked as answer]
A `PreParrsedDocumentEntry` is really just an API wrapper class represent EITHER a Document or errors with one

You only need this when your provider calls back to the graphql-java

You can serialise the `Document` as it implements `Serializable`

So you would

* derive a key say from the query text
* look up in cache
* if you get a miss, then do this
* parse and validate the ExecutionInput via the `parseAndValidateFunction` method first
* if its all valid, then grab the Document and then serialise that to cache
* derive a key say from the query text
* return the result

On the next query, you would

* derive a key say from the query text
* look up in cache
* if you get a hit, de-serialse the document
* return a new `PreParrsedDocumentEntry` containing that de-serialsed `Document`


The idea behind [PreparsedDocumentProvider](https://www.javadoc.io/doc/com.graphql-java/graphql-java/18.0/graphql/execution/preparsed/PreparsedDocumentProvider.html) is to avoid having the parse every document every request.

**Reply by softprops** [reply]
excellent answer. thank you! 

I'm currently using the [PreparsedDocumentProvider](https://www.javadoc.io/doc/com.graphql-java/graphql-java/18.0/graphql/execution/preparsed/PreparsedDocumentProvider.html) with an in memory cache and it works terrifically. 

Im very interested in extending that to a cross process redis cache. 

Thank you for maintaining such a well thought out library.
